### PR TITLE
Reintroduce recording of function targets

### PR DESCRIPTION
**Pull request by o-**
based on top of #181

### CODE REVIEW

File: rir/src/ir/Compiler.cpp [hunk 1]
```diff
@@ -1077,10 +1077,11 @@ void compileCall(Context& ctx, SEXP ast, SEXP fun, SEXP args) {
     }
     assert(callArgs.size() < BC::MAX_NUM_ARGS);
 
-    if (hasNames)
+    if (hasNames) {
```
Comment (mhyee) [inline rir/src/ir/Compiler.cpp:1080]: Do we have an agreed convention for braces? I've seen braced and braceless if-statements, but never commented on them. (I prefer always having braces.)

Unfortunately, it looks like clang-format can't enforce this.

File: rir/src/ir/RuntimeFeedback.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+namespace rir {
+
+#pragma pack(push)
+#pragma pack(1)
+
+struct CallFeedback {
+    static constexpr unsigned CounterBits = 30;
+    static constexpr unsigned CounterOverflow = (1 << CounterBits) - 1;
+    static constexpr unsigned TargetBits = 2;
+    static constexpr unsigned MaxTargets = (1 << TargetBits) - 1;
```
Comment (mhyee) [inline rir/src/ir/RuntimeFeedback.h:16]: I think we're off-by-one here... with 2 bits, we can represent 4 targets. But then the condition below (line 25) will need to be more complicated...

Actually, the problem is on line 31, where we do not want to increment `numTargets` if its value is 4 (assuming we have 2 target bits).

(Or maybe I'm completely wrong. I hate bit twiddling and off-by-one errors.)

File: rir/src/ir/RuntimeFeedback.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            for (; i < numTargets; ++i)
+                if (targets[i] == callee)
+                    break;
+            if (i == numTargets)
+                targets[numTargets++] = callee;
+        }
+    }
+};
+
+static_assert(sizeof(CallFeedback) == 7 * sizeof(uint32_t),
```
Comment (mhyee) [inline rir/src/ir/RuntimeFeedback.h:36]: Add a comment for where `7` comes from, in case it needs to be modified?

File: rir/src/ir/insns.h [hunk 4]
```diff
@@ -438,4 +438,9 @@ DEF_INSTR(ret_, 0, 1, 0, 1)
  */
 DEF_INSTR(int3_, 0, 0, 0, 1)
 
+/*
+ * recording bytecodes are used to collect information
```
Comment (mhyee) [inline rir/src/ir/insns.h:442]: Add some documentation about how this instruction connects to `CallFeedback`? Specifically, it's a bit weird seeing that this single instruction needs 7 immediate args.

File: rir/src/ir/RuntimeFeedback.h (hunk 2 above)
Comment (mhyee) [reply rir/src/ir/RuntimeFeedback.h:16]: Yeah, I think there's some conflation in how `numTargets` is being used as an index into the array, and so `MaxTargets` is actually `MaxTargetIndex`.

**Review by mhyee — COMMENTED** [review]
I don't really remember how the old target recording worked, but from my first impressions, this approach looks pretty good.

I guess my question is how frequently will `record_call_` be emitted? Seven immediates seems like quite a bit. And if start recording types, that's also more stuff we're shoving into the bytecode stream.

Then again, this is probably simpler than keeping stuff separate from the bytecode stream. (I guess that's what we did before?)

File: rir/src/ir/RuntimeFeedback.h (hunk 2 above)
Comment (o-) [reply rir/src/ir/RuntimeFeedback.h:16]: I am not sure if I understand. I think it's actually correct. In this concrete case we have 2 target bits, so MaxTargets is 3.  If we find a target, then its stored at targets[0] and numTargets is increased to 1. If we find a second target, we store at 1 and increase to 2. Finally if we find a third target, we store at 2 and increase to MaxTargets. From now on we will no longer store more targets. What we lost, is the information if we had exactly 3 targets or more. We could steal one bit from the counter, if we think it is important to know that..

File: rir/src/ir/RuntimeFeedback.h (hunk 3 above)
Comment (o-) [reply rir/src/ir/RuntimeFeedback.h:36]: i documented it in the other PR no

**Comment by o-** [comment]
hey, I agree with the general sentiment, that its kind of expensive to store this in the BC stream... on the other hand it is so much more convenient. We used to have it dangling at the end of the Code objects and it was a mess. another nice thing is: if we don't do it this way (with dedicated instructions) then every instruction which potentially wants to record feedback, needs a bit telling it, if it should record or not (because presumably we only want to record when it makes sense...)

File: rir/src/ir/RuntimeFeedback.h (hunk 2 above)
Comment (mhyee) [reply rir/src/ir/RuntimeFeedback.h:16]: So I think the code works exactly as you described. When I first looked at this code, my thought was "Well, with 2 bits to record the target, we should be able to store 4 targets, but right now you're only storing up to 3."

After thinking some more, I think it's fine the way it is (but maybe with some explanation so I don't run into this and get confused again). Storing one more target probably isn't super helpful, and the logic would become far more complicated. (Being able to index "one past the last element" is so useful...)

**Comment by mhyee** [comment]
> if we don't do it this way (with dedicated instructions) then every instruction which potentially wants to record feedback, needs a bit telling it, if it should record or not (because presumably we only want to record when it makes sense...)

Yeah, having the dedicated instructions seems like a better approach to me.

**Pull request closed without merging**